crates/hir_ty/src/diagnostics/match_check/usefulness.rs

```diff
@@ -0,0 +1,1180 @@
+        // FIXME(iDawer) implement exhaustive_patterns feature. More info in:
+        // Tracking issue for RFC 1872: exhaustive_patterns feature https://github.com/rust-lang/rust/issues/51085
+        false
+    }
+
+    /// Returns whether the given type is an enum from another crate declared `#[non_exhaustive]`.
+    pub(super) fn is_foreign_non_exhaustive_enum(&self, enum_id: hir_def::EnumId) -> bool {
+        let has_non_exhaustive_attr =
+            self.db.attrs(enum_id.into()).by_key("non_exhaustive").exists();
+        let is_local =
+            hir_def::AdtId::from(enum_id).module(self.db.upcast()).krate() == self.module.krate();
+        has_non_exhaustive_attr && !is_local
+    }
+
+    // Rust feature described as "Allows exhaustive pattern matching on types that contain uninhabited types."
+    pub(super) fn feature_exhaustive_patterns(&self) -> bool {
+        // FIXME see MatchCheckCtx::is_uninhabited
+        false
+    }
+
+    pub(super) fn alloc_pat(&self, pat: Pat) -> PatId {
+        self.pattern_arena.borrow_mut().alloc(pat)
+    }
+
+    /// Get type of a pattern. Handles expanded patterns.
+    pub(super) fn type_of(&self, pat: PatId) -> Ty {
+        self.pattern_arena.borrow()[pat].ty.clone()
+    }
+}
+
+#[derive(Copy, Clone)]
+pub(super) struct PatCtxt<'a> {
+    pub(super) cx: &'a MatchCheckCtx<'a>,
+    /// Type of the current column under investigation.
+    pub(super) ty: &'a Ty,
+    /// Whether the current pattern is the whole pattern as found in a match arm, or if it's a
+    /// subpattern.
+    pub(super) is_top_level: bool,
+}
+
+pub(crate) fn expand_pattern(pat: Pat) -> Pat {
+    LiteralExpander.fold_pattern(&pat)
+}
+
+struct LiteralExpander;
+
+impl PatternFolder for LiteralExpander {
+    fn fold_pattern(&mut self, pat: &Pat) -> Pat {
+        match (pat.ty.kind(&Interner), pat.kind.as_ref()) {
+            (_, PatKind::Binding { subpattern: Some(s), .. }) => s.fold_with(self),
+            _ => pat.super_fold_with(self),
+        }
+    }
+}
+
+impl Pat {
+    fn _is_wildcard(&self) -> bool {
+        matches!(*self.kind, PatKind::Binding { subpattern: None, .. } | PatKind::Wild)
+    }
+}
+
+impl PatIdExt for PatId {
+    fn is_or_pat(self, cx: &MatchCheckCtx<'_>) -> bool {
+        matches!(*cx.pattern_arena.borrow()[self].kind, PatKind::Or { .. })
+    }
+
+    /// Recursively expand this pattern into its subpatterns. Only useful for or-patterns.
+    fn expand_or_pat(self, cx: &MatchCheckCtx<'_>) -> Vec<Self> {
+        fn expand(pat: PatId, vec: &mut Vec<PatId>, pat_arena: &mut PatternArena) {
+            if let PatKind::Or { pats } = pat_arena[pat].kind.as_ref() {
+                let pats = pats.clone();
+                for pat in pats {
+                    // FIXME(iDawer): Ugh, I want to go back to references (PatId -> &Pat)
+                    let pat = pat_arena.alloc(pat.clone());
+                    expand(pat, vec, pat_arena);
+                }
```

Comment:
I used indexed arena which brought some extra clones here and in `deconstruct_pat::Fields::replace_with_pattern_arguments(..)`.

Reply:
Yeah, I wouldn't want about cloning at all at this stage. 

Though, I am curios, why we use *both* arena and pointer indirection?

I'd expect us to use only one appraoch:

* Either `Pat` stores `Box<Pat>` inside, and there's no arenas anywhere
* Or `Pat` stores `PatId` inside, and all patterns are inside an arena

But, again, I think this is something we can adjust as we go.

Reply:
>     * Either `Pat` stores `Box<Pat>` inside, and there's no arenas anywhere

The algo constructs intermediate and witness patterns using parts of existing ones anyway so deep cloning will be inevitable. I didn't track ownership down closely, cannot say for sure.

>     * Or `Pat` stores `PatId` inside, and all patterns are inside an arena

Hmm, it makes sense. Even better all `Pat`s store `PatId` inside, this would allow replacing deep clones `Vec<Pat>::clone()` with forgivable `[PatId]::clone()`.

I can't see how to completely remove clones here without resorting to `typed-arena` dependency.
What do you think about adding `typed-arena` dep? It gives references `<TypedArena<T>>::alloc(&self, value: T) -> &mut T` and would allow to switch to use `&Pat` and remove dancing with iterator invalidation.

I'll point to this discussion in fixme.